core/repository/sparql/src/test/java/org/eclipse/rdf4j/repository/sparql/SPARQLConnectionTest.java

```diff
@@ -48,4 +53,52 @@ public void commitOnEmptyTxnDoesNothing() throws Exception {
 		verify(client, never()).sendUpdate(any(), any(), any(), any(), anyBoolean(), anyInt(), any());
 		verify(client, never()).sendUpdate(any(), any(), any(), any(), anyBoolean(), any());
 	}
+
+	@Test
+	public void testGroupingAddsInInsert() throws Exception {
+		ArgumentCaptor<String> sparqlUpdateCaptor = ArgumentCaptor.forClass(String.class);
+
+		subject.begin();
+		subject.add(FOAF.PERSON, RDF.TYPE, RDFS.CLASS);
+		subject.add(FOAF.AGENT, RDF.TYPE, RDFS.CLASS);
+		subject.commit();
+
+		verify(client).sendUpdate(any(), sparqlUpdateCaptor.capture(), any(), any(), anyBoolean(), anyInt(), any());
+
+		String sparqlUpdate = sparqlUpdateCaptor.getValue();
+		String expectedTriple1 = "<" + FOAF.PERSON + "> <" + RDF.TYPE + "> <" + RDFS.CLASS + ">";
+		String expectedTriple2 = "<" + FOAF.AGENT + "> <" + RDF.TYPE + "> <" + RDFS.CLASS + ">";
+
+		assertThat(sparqlUpdate).containsOnlyOnce("INSERT DATA").contains(expectedTriple1).contains(expectedTriple2);
+	}
+
+	@Test
+	public void testHandlingAddsRemoves() throws Exception {
+		ArgumentCaptor<String> sparqlUpdateCaptor = ArgumentCaptor.forClass(String.class);
+
+		subject.begin();
+		subject.add(FOAF.PERSON, RDF.TYPE, RDFS.CLASS);
+		subject.add(FOAF.AGENT, RDF.TYPE, RDFS.CLASS);
+		subject.remove(FOAF.BIRTHDAY, RDF.TYPE, RDF.PROPERTY);
+		subject.add(FOAF.AGE, RDF.TYPE, RDF.PROPERTY);
+		subject.commit();
+
+		verify(client).sendUpdate(any(), sparqlUpdateCaptor.capture(), any(), any(), anyBoolean(), anyInt(), any());
+
+		String sparqlUpdate = sparqlUpdateCaptor.getValue();
+
+		String expectedAddedTriple1 = "<" + FOAF.PERSON + "> <" + RDF.TYPE + "> <" + RDFS.CLASS + "> .";
+		String expectedAddedTriple2 = "<" + FOAF.AGENT + "> <" + RDF.TYPE + "> <" + RDFS.CLASS + "> .";
+		String expectedAddedTriple3 = "<" + FOAF.AGE + "> <" + RDF.TYPE + "> <" + RDF.PROPERTY + "> ";
+		String expectedRemovedTriple1 = "<" + FOAF.BIRTHDAY + "> <" + RDF.TYPE + "> <" + RDF.PROPERTY + "> .";
+
+		String expectedSequence = "INSERT DATA[^{]*\\{[^}]*\\}[^D]+DELETE DATA[^{]*\\{[^}]*\\}[^I]+INSERT DATA.*";
+
+		assertThat(sparqlUpdate).containsPattern(expectedSequence);
+		assertThat(sparqlUpdate).contains(expectedAddedTriple1)
+				.contains(expectedAddedTriple2)
+				.contains(expectedAddedTriple3)
+				.contains(expectedRemovedTriple1);
+
+	}
```

Comment:
Good we have some tests. Do we have any integration tests that would cover this too? Benchmarks would also be great, so we don't have a regression of this some time in the future.

Reply:
Yes we have quite extensive integration tests for this, see rdf4j-repository-compliance. 

I don't think we have benchmarks though. I'll see what I can do this weekend. But I must admit I suck at benchmarks.